ICE: `assertion failed: bpos.to_u32() >= mbc.pos.to_u32() + mbc.bytes as u32`

<!--
[31mICE[0m: Rustc ./ice.rs '' 'thread 'rustc' panicked at compiler/rustc_span/src/lib.rs:2028:17: 'assertion failed: bpos.to_u32() >= mbc.pos.to_u32() + mbc.bytes as u32'', 'thread 'rustc' panicked at compiler/rustc_span/src/lib.rs:2028:17: 'assertion failed: bpos.to_u32() >= mbc.pos.to_u32() + mbc.bytes as u32''
File: /tmp/im/redu/ice.rs
-->

snippet:
````rust
pub fn main() {
    main （arr[i]）;
}


````

Version information
````
rustc 1.82.0-nightly (b586701f7 2024-08-05)
binary: rustc
commit-hash: b586701f78a6d5c7f618b76e7ae3cace9a6fbf37
commit-date: 2024-08-05
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error: unknown start of token: \u{ff08}
 --> /tmp/icemaker_global_tempdir.hijyp6Vy7Iut/rustc_testrunner_tmpdir_reporting.2jS84jyEH0Ar/mvce.rs:2:10
  |
2 |     main （arr[i]）;
  |          ^^
  |
help: Unicode character '（' (Fullwidth Left Parenthesis) looks like '(' (Left Parenthesis), but it is not
  |
2 |     main (arr[i]）;
  |          ~

error: unknown start of token: \u{ff09}
 --> /tmp/icemaker_global_tempdir.hijyp6Vy7Iut/rustc_testrunner_tmpdir_reporting.2jS84jyEH0Ar/mvce.rs:2:17
  |
2 |     main （arr[i]）;
  |                  ^^
  |
help: Unicode character '）' (Fullwidth Right Parenthesis) looks like ')' (Right Parenthesis), but it is not
  |
2 |     main （arr[i]);
  |                  ~

error[E0425]: cannot find value `arr` in this scope
   --> /tmp/icemaker_global_tempdir.hijyp6Vy7Iut/rustc_testrunner_tmpdir_reporting.2jS84jyEH0Ar/mvce.rs:2:11
    |
2   |     main （arr[i]）;
    |            ^^^ help: a tuple variant with a similar name exists: `Err`
    |
   ::: /home/matthias/.rustup/toolchains/master/lib/rustlib/src/rust/library/core/src/result.rs:536:5
    |
536 |     Err(#[stable(feature = "rust1", since = "1.0.0")] E),
    |     --- similarly named tuple variant `Err` defined here

error[E0425]: cannot find value `i` in this scope
 --> /tmp/icemaker_global_tempdir.hijyp6Vy7Iut/rustc_testrunner_tmpdir_reporting.2jS84jyEH0Ar/mvce.rs:2:15
  |
2 |     main （arr[i]）;
  |                ^ not found in this scope

error[E0061]: this function takes 0 arguments but 1 argument was supplied
 --> /tmp/icemaker_global_tempdir.hijyp6Vy7Iut/rustc_testrunner_tmpdir_reporting.2jS84jyEH0Ar/mvce.rs:2:5
  |
2 |     main （arr[i]）;
  |     ^^^^   ------ unexpected argument
  |
note: function defined here
 --> /tmp/icemaker_global_tempdir.hijyp6Vy7Iut/rustc_testrunner_tmpdir_reporting.2jS84jyEH0Ar/mvce.rs:1:8
  |
1 | pub fn main() {
  |        ^^^^
thread 'rustc' panicked at compiler/rustc_span/src/lib.rs:2028:17:
assertion failed: bpos.to_u32() >= mbc.pos.to_u32() + mbc.bytes as u32
stack backtrace:
   0:     0x7379e402ef5d - std::backtrace_rs::backtrace::libunwind::trace::h54121657fad6b938
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x7379e402ef5d - std::backtrace_rs::backtrace::trace_unsynchronized::hd2e82ea9904f2ea9
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7379e402ef5d - std::sys::backtrace::_print_fmt::h75adf07dfa40e5d2
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/sys/backtrace.rs:66:9
   3:     0x7379e402ef5d - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h98e5ee4b224fe254
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/sys/backtrace.rs:39:26
   4:     0x7379e407f6eb - core::fmt::rt::Argument::fmt::h0835d2bcbbdcecae
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/core/src/fmt/rt.rs:173:76
   5:     0x7379e407f6eb - core::fmt::write::h41425652aafb8d55
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/core/src/fmt/mod.rs:1178:21
   6:     0x7379e4022cc3 - std::io::Write::write_fmt::h722492c13c12a6df
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/io/mod.rs:1823:15
   7:     0x7379e4031752 - std::sys::backtrace::BacktraceLock::print::hcfe3c665d987f44a
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/sys/backtrace.rs:42:9
   8:     0x7379e4031752 - std::panicking::default_hook::{{closure}}::h73b850e758a43b44
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/panicking.rs:266:22
   9:     0x7379e40313be - std::panicking::default_hook::h0ed5c4bdc6aad088
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/panicking.rs:293:9
  10:     0x7379e0498bf7 - std[5068228d6f0dc02a]::panicking::update_hook::<alloc[8fcb41ef64dc878b]::boxed::Box<rustc_driver_impl[b43ed4103cc78677]::install_ice_hook::{closure#0}>>::{closure#0}
  11:     0x7379e4032142 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h88c4c4826ca7daef
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/alloc/src/boxed.rs:2164:9
  12:     0x7379e4032142 - std::panicking::rust_panic_with_hook::h9849dc0f6b06d65d
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/panicking.rs:805:13
  13:     0x7379e4031dc3 - std::panicking::begin_panic_handler::{{closure}}::hf77d3bd3ba796e45
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/panicking.rs:664:13
  14:     0x7379e402f449 - std::sys::backtrace::__rust_end_short_backtrace::h61e613d4057be1a3
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/sys/backtrace.rs:170:18
  15:     0x7379e4031a84 - rust_begin_unwind
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/panicking.rs:662:5
  16:     0x7379e407bd43 - core::panicking::panic_fmt::hc58ff999a1ba55da
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/core/src/panicking.rs:74:14
  17:     0x7379e407bdcc - core::panicking::panic::hbce99ce8443817ed
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/core/src/panicking.rs:148:5
  18:     0x7379de9bb1c0 - <rustc_span[6b6bb1cd45519793]::source_map::SourceMap>::lookup_char_pos
  19:     0x7379e2aafbff - <rustc_span[6b6bb1cd45519793]::source_map::SourceMap>::is_valid_span
  20:     0x7379e2aaaef7 - <core[430efa8e2f06c6b]::iter::adapters::filter_map::FilterMap<core[430efa8e2f06c6b]::iter::adapters::cloned::Cloned<core[430efa8e2f06c6b]::iter::adapters::filter::Filter<core[430efa8e2f06c6b]::slice::iter::Iter<rustc_errors[1c74d9fcab0bdd42]::Substitution>, <rustc_errors[1c74d9fcab0bdd42]::CodeSuggestion>::splice_lines::{closure#0}>>, <rustc_errors[1c74d9fcab0bdd42]::CodeSuggestion>::splice_lines::{closure#1}> as core[430efa8e2f06c6b]::iter::traits::iterator::Iterator>::next
  21:     0x7379e2aa67b2 - <rustc_errors[1c74d9fcab0bdd42]::emitter::HumanEmitter>::emit_messages_default
  22:     0x7379e2ba1227 - <rustc_errors[1c74d9fcab0bdd42]::emitter::HumanEmitter as rustc_errors[1c74d9fcab0bdd42]::emitter::Emitter>::emit_diagnostic
  23:     0x7379e2ba27ef - <rustc_errors[1c74d9fcab0bdd42]::DiagCtxtInner>::emit_diagnostic::{closure#3}
  24:     0x7379e2b9fd51 - rustc_interface[dcf23a6b0366e6de]::callbacks::track_diagnostic::<core[430efa8e2f06c6b]::option::Option<rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>>
  25:     0x7379e2b9e1fe - <rustc_errors[1c74d9fcab0bdd42]::DiagCtxtInner>::emit_diagnostic
  26:     0x7379e2b9e0ad - <rustc_errors[1c74d9fcab0bdd42]::DiagCtxtHandle>::emit_diagnostic
  27:     0x7379e2b9d255 - <rustc_span[6b6bb1cd45519793]::ErrorGuaranteed as rustc_errors[1c74d9fcab0bdd42]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  28:     0x7379e07ac5db - <rustc_hir_typeck[3b9cefa2bb200429]::fn_ctxt::FnCtxt>::report_arg_errors
  29:     0x7379deef06b5 - <rustc_hir_typeck[3b9cefa2bb200429]::fn_ctxt::FnCtxt>::confirm_builtin_call
  30:     0x7379e23ac340 - <rustc_hir_typeck[3b9cefa2bb200429]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  31:     0x7379e23a5633 - <rustc_hir_typeck[3b9cefa2bb200429]::fn_ctxt::FnCtxt>::check_block_with_expected
  32:     0x7379e23acf98 - <rustc_hir_typeck[3b9cefa2bb200429]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  33:     0x7379e239089c - rustc_hir_typeck[3b9cefa2bb200429]::check::check_fn
  34:     0x7379e2369837 - rustc_hir_typeck[3b9cefa2bb200429]::typeck
  35:     0x7379e23691f3 - rustc_query_impl[bae30e77a033ddc]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[bae30e77a033ddc]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[6ed3e96e78853911]::query::erase::Erased<[u8; 8usize]>>
  36:     0x7379e1d92ab5 - rustc_query_system[9eda3ab91ed39e2e]::query::plumbing::try_execute_query::<rustc_query_impl[bae30e77a033ddc]::DynamicConfig<rustc_query_system[9eda3ab91ed39e2e]::query::caches::VecCache<rustc_span[6b6bb1cd45519793]::def_id::LocalDefId, rustc_middle[6ed3e96e78853911]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[bae30e77a033ddc]::plumbing::QueryCtxt, false>
  37:     0x7379e1d91ca3 - rustc_query_impl[bae30e77a033ddc]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  38:     0x7379e1d918c0 - <rustc_middle[6ed3e96e78853911]::hir::map::Map>::par_body_owners::<rustc_hir_analysis[c02afad8edaa9fa]::check_crate::{closure#4}>::{closure#0}
  39:     0x7379e1d8f6ac - rustc_hir_analysis[c02afad8edaa9fa]::check_crate
  40:     0x7379e1d85055 - rustc_interface[dcf23a6b0366e6de]::passes::analysis
  41:     0x7379e1d84c07 - rustc_query_impl[bae30e77a033ddc]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[bae30e77a033ddc]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[6ed3e96e78853911]::query::erase::Erased<[u8; 1usize]>>
  42:     0x7379e29a816e - rustc_query_system[9eda3ab91ed39e2e]::query::plumbing::try_execute_query::<rustc_query_impl[bae30e77a033ddc]::DynamicConfig<rustc_query_system[9eda3ab91ed39e2e]::query::caches::SingleCache<rustc_middle[6ed3e96e78853911]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[bae30e77a033ddc]::plumbing::QueryCtxt, false>
  43:     0x7379e29a7ecf - rustc_query_impl[bae30e77a033ddc]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  44:     0x7379e28adc23 - rustc_interface[dcf23a6b0366e6de]::interface::run_compiler::<core[430efa8e2f06c6b]::result::Result<(), rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>, rustc_driver_impl[b43ed4103cc78677]::run_compiler::{closure#0}>::{closure#1}
  45:     0x7379e27e9709 - std[5068228d6f0dc02a]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[dcf23a6b0366e6de]::util::run_in_thread_with_globals<rustc_interface[dcf23a6b0366e6de]::util::run_in_thread_pool_with_globals<rustc_interface[dcf23a6b0366e6de]::interface::run_compiler<core[430efa8e2f06c6b]::result::Result<(), rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>, rustc_driver_impl[b43ed4103cc78677]::run_compiler::{closure#0}>::{closure#1}, core[430efa8e2f06c6b]::result::Result<(), rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>>::{closure#0}, core[430efa8e2f06c6b]::result::Result<(), rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[430efa8e2f06c6b]::result::Result<(), rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>>
  46:     0x7379e27e94b2 - <<std[5068228d6f0dc02a]::thread::Builder>::spawn_unchecked_<rustc_interface[dcf23a6b0366e6de]::util::run_in_thread_with_globals<rustc_interface[dcf23a6b0366e6de]::util::run_in_thread_pool_with_globals<rustc_interface[dcf23a6b0366e6de]::interface::run_compiler<core[430efa8e2f06c6b]::result::Result<(), rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>, rustc_driver_impl[b43ed4103cc78677]::run_compiler::{closure#0}>::{closure#1}, core[430efa8e2f06c6b]::result::Result<(), rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>>::{closure#0}, core[430efa8e2f06c6b]::result::Result<(), rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[430efa8e2f06c6b]::result::Result<(), rustc_span[6b6bb1cd45519793]::ErrorGuaranteed>>::{closure#1} as core[430efa8e2f06c6b]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  47:     0x7379e403be5b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h47908fbe26776c73
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/alloc/src/boxed.rs:2150:9
  48:     0x7379e403be5b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h54da00d4b6e8da22
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/alloc/src/boxed.rs:2150:9
  49:     0x7379e403be5b - std::sys::pal::unix::thread::Thread::new::thread_start::hbf56854fd7d508a2
                               at /rustc/b586701f78a6d5c7f618b76e7ae3cace9a6fbf37/library/std/src/sys/pal/unix/thread.rs:105:17
  50:     0x7379dcea6ded - <unknown>
  51:     0x7379dcf2a0dc - <unknown>
  52:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.82.0-nightly (b586701f7 2024-08-05) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `main`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0061, E0425.
For more information about an error, try `rustc --explain E0061`.

```

</p>
</details>

<!--
query stack:
536 |     Err(#[stable(feature = "rust1", since = "1.0.0")] E),
#0 [typeck] type-checking `main`
#1 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
@@ -22,7 +22,7 @@ use rustc_middle::ty::{self, IsSuggestable, Ty, TyCtxt};
 use rustc_middle::{bug, span_bug};
 use rustc_session::Session;
 use rustc_span::symbol::{kw, Ident};
-use rustc_span::{sym, BytePos, Span, DUMMY_SP};
+use rustc_span::{sym, Span, DUMMY_SP};
 use rustc_trait_selection::error_reporting::infer::{FailureCode, ObligationCauseExt};
 use rustc_trait_selection::infer::InferCtxtExt;
 use rustc_trait_selection::traits::{self, ObligationCauseCode, SelectionContext};
@@ -1140,8 +1140,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 .get(arg_idx + 1)
                                 .map(|&(_, sp)| sp)
                                 .unwrap_or_else(|| {
-                                    // Subtract one to move before `)`
-                                    call_expr.span.with_lo(call_expr.span.hi() - BytePos(1))
+                                    // Try to move before `)`. Note that `)` here is not necessarily
+                                    // the latin right paren, it could be a Unicode-confusable that
+                                    // looks like a `)`, so we must not use `- BytePos(1)`
+                                    // manipulations here.
+                                    self.tcx().sess.source_map().end_point(call_expr.span)
                                 });
 
                             // Include next comma

--- a/tests/ui/typeck/suggest-arg-comma-delete-ice.rs
+++ b/tests/ui/typeck/suggest-arg-comma-delete-ice.rs
@@ -1,0 +1,19 @@
+//! Previously, we tried to remove extra arg commas when providing extra arg removal suggestions.
+//! One of the edge cases is having to account for an arg that has a closing delimiter `)`
+//! following it. However, the previous suggestion code assumed that the delimiter is in fact
+//! exactly the 1-byte `)` character. This assumption was proven incorrect, because we recover
+//! from Unicode-confusable delimiters in the parser, which means that the ending delimiter could be
+//! a multi-byte codepoint that looks *like* a `)`. Subtracing 1 byte could land us in the middle of
+//! a codepoint, triggering a codepoint boundary assertion.
+//!
+//! issue: rust-lang/rust#128717
+
+fn main() {
+    // The following example has been modified from #128717 to remove irrelevant Unicode as they do
+    // not otherwise partake in the right delimiter calculation causing the codepoint boundary
+    // assertion.
+    main(rahh）;
+    //~^ ERROR unknown start of token
+    //~| ERROR this function takes 0 arguments but 1 argument was supplied
+    //~| ERROR cannot find value `rahh` in this scope
+}

--- a/tests/ui/typeck/suggest-arg-comma-delete-ice.stderr
+++ b/tests/ui/typeck/suggest-arg-comma-delete-ice.stderr
@@ -1,0 +1,38 @@
+error: unknown start of token: \u{ff09}
+  --> $DIR/suggest-arg-comma-delete-ice.rs:15:14
+   |
+LL |     main(rahh）;
+   |              ^^
+   |
+help: Unicode character '）' (Fullwidth Right Parenthesis) looks like ')' (Right Parenthesis), but it is not
+   |
+LL |     main(rahh);
+   |              ~
+
+error[E0425]: cannot find value `rahh` in this scope
+  --> $DIR/suggest-arg-comma-delete-ice.rs:15:10
+   |
+LL |     main(rahh）;
+   |          ^^^^ not found in this scope
+
+error[E0061]: this function takes 0 arguments but 1 argument was supplied
+  --> $DIR/suggest-arg-comma-delete-ice.rs:15:5
+   |
+LL |     main(rahh）;
+   |     ^^^^ ---- unexpected argument
+   |
+note: function defined here
+  --> $DIR/suggest-arg-comma-delete-ice.rs:11:4
+   |
+LL | fn main() {
+   |    ^^^^
+help: remove the extra argument
+   |
+LL -     main(rahh）;
+LL +     main(）;
+   |
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0061, E0425.
+For more information about an error, try `rustc --explain E0061`.

